jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -349,92 +354,57 @@ private long getAvailableMemory(boolean direct)
     {
         for (RetainedBucket bucket : poolArray)
-        {
-            bucket.getPool().stream().forEach(entry ->
-            {
-                if (entry.remove())
-                {
-                    RetainableByteBuffer pooled = entry.getPooled();
-                    // Calling getPooled can return null if the entry was not yet enabled.
-                    if (pooled != null)
-                    {
-                        memoryCounter.addAndGet(-pooled.capacity());
-                        removed(pooled);
-                    }
-                }
-            });
-        }
+            bucket.getPool().stream().forEach(Pool.Entry::remove);
     }
 
     private void releaseExcessMemory(boolean direct)
     {
         long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
-        if (maxMemory > 0)
-        {
-            long excess = getMemory(direct) - maxMemory;
-            if (excess > 0)
-                evict(direct, excess);
-        }
-    }
+        if (maxMemory <= 0)
+            return;
 
-    /**
-     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
-     *
-     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
-     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
-     */
-    private void evict(boolean direct, long excess)
-    {
-        if (LOG.isDebugEnabled())
-            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
-        long now = NanoTime.now();
-        long totalClearedCapacity = 0L;
+        if (_evictor.compareAndSet(false, true))
+        {
+            // we are the one and only thread evicting
+            try
+            {
+                RetainedBucket[] buckets = direct ? _direct : _indirect;
 
-        RetainedBucket[] buckets = direct ? _direct : _indirect;
+                long excess = getMemory(direct) - maxMemory;
+                if (excess > 0)
+                {
+                    // find the bucket with the most idle memory
+                    RetainedBucket mostIdleBucket = null;
+                    long mostIdleSize = -1;
+                    for (RetainedBucket bucket : buckets)
+                    {
+                        long idle = (long)bucket.getPool().getIdleCount() * bucket._capacity;
+                        if (idle > mostIdleSize)
+                        {
+                            mostIdleBucket = bucket;
+                            mostIdleSize = idle;
+                        }
+                    }
 
-        while (totalClearedCapacity < excess)
-        {
-            // Run through all the buckets to avoid removing
-            // the buffers only from the first bucket(s).
-            for (RetainedBucket bucket : buckets)
+                    if (mostIdleBucket != null)
+                        mostIdleBucket.evict(excess);
```

Comment:
The call to `getMemory()` returns mixed information, because it assumed in-use entries from `ConcurrentPool` as consuming memory (they are not), and for `QueuedPool` the entries in use are not present in the queue, so they are not accounted for.

The `for` loop may find `mostIdleSize=0`, as the pool is under pressure and no bucket is idle.
So it defaults to the first bucket.

Then it proceeds to evict possibly _all_ the entries from the smallest bucket, which is actually the one that has no idles and therefore is the most used one, thrashing it because now all the entries must be re-allocated.

Finally, it may fail to keep the memory under the limit and has to set the oversize flag, which was the whole point of this feature (keep the memory constrained).

Isn't the oversize allowed to grow indefinitely?
Yes, because more and more ConcurrentEntries can be request and be in use, but counted as used memory, and this method may never find an idle bucket, always hitting the first, now empty, bucket, therefore never reducing the memory.

Reply:
@sbordet if no bucket is idle, then nothing will be evicted, because the algorithm only evicts idle entries - not all entries.  I don't know why you keep saying it evicts all entries from a bucket.  It will only evict idle entries.

However, you are correct about `getMemory`, as it should only count idle buffers and not all entries, so that will make it a bit more expensive.  The current pool also makes this mistake and is counting all entries in the pool, idle or otherwise.

But you are wrong about it not being able to reduce memory.   It will only fail to reduce memory if there are no idle entries... which is essentially an empty pool (by your updated definition of pool size), so nothing needs to be evicted anyway.

The oversize flag is just to address the concern of @lorban that one thread might end up doing all the work evicting.  So now a thread will do a single pass and if that is incomplete, the oversize flag is set so that the next release will do another pass.  It is just spreading the work over multiple threads.   I actually think that a single thread should be able to iterate on the evictions.

Reply:
I missed the filter on `isIdle()` in `evict()`.

However, that is in a race with the subsequent removal.
The removal does take into account that even if `remove()` returns `false` (because it is in use), it will be _eventually_ removed, so its memory should be accounted for anyway.
Also, I don't see leaks being handled properly, unless you switch to the new semantic of `getMemory()`.

This PR cannot precisely count the memory by looking for idle buffers, as you will be in a race.
So the feature of memory constrain is broken.

Furthermore, it will always try to remove from 1 bucket only, which would leave possibly many idle buffers in smaller buckets, just because there is 1 buffer idle in a bigger bucket; e.g. leave 31 idle 1KiB buffers to remove 1 32KiB buffer that by the time it's attempted to be evicted may be in use, and therefore not evicted (because it will be filtered out by the filter I missed).

Reply:
@sbordet I've switch to accounting memory only for idle buffers.
I avoid the race now by doing an acquire, which will only return an idle buffer, which I then remove.

I agree that it might be sub-optimal to remove from just one bucket.  But the win is that the most idle bucket does not need to be recalculated.

Perhaps the algorithm should be to pick the bucket with the large number of idle entries, rather than just the largest memory usage.   Perhaps just picking a random bucket would work.. or cycling through all the buckets.

Reply:
@gregw note also that you this PR is counting over the entries in the buckets multiple times.

Given that `QueuedPool` by default has `Integer.MAX_VALUE` capacity, a load spike may create a lot of entries and counting through them for filtering, for calculating sizes that are in a race with usages, etc. it is exactly the cost that now is causing CPU usage issues.

This PR still leaves buckets that are full of idle buffers in case of "adapts to changes in size frequency" possibly hitting exactly only on the largest bucket, it will consume a lot of CPU in worst cases due to running over possibly millions of entries (again, the worst case) to check whether they are idle, it is not able to respect maxMemory, all the checks for idleness and choice of buckets are racy, making it eventually impossible to understand bad behaviors that we may have not discovered yet.

Reply:
@sbordet @lorban If we are above max size and have a choice of evicting 1 32MB buffer or 32 1MB buffers, then which should we do (assuming no change in size frequency)?

I think the 32MB buffer as that allows 32 subsequent acquires to find a pooled buffer, whilst kicking out the 1MB buffers might require 32 new reservations to satisfy future demand.

But if we only need 16MB of space, is it better to kick out 16 x 1MB or 1 x 32MB?   That's a tough question.   It feels wrong to kick out the 32MB as that is throwing out extra memory, but then it will be cheaper to replace 1 32MB buffer than to replace 16x1MB buffers.

My current algorithm to evict from the bucket with most idle space, is really neither of these.   Perhaps we should evict from the bucket with most idle entries and not consider the size?  Or least recently used bucket? 

The other issue, is that if there is only 1 size bucket in use, the @sbordet PR will be a lot better, as it never adds a buffer that will then need to be evicted.   This PR will release, acquire, remove.    I think 1 size bucket is probably frequent.... but are we often hitting the max capacity of the pool?

I'm currently thinking we need an algorithm that can work both ways: avoiding adding a buffer and looking for other buckets that can be evicted.   I have some ideas... stand by.

Reply:
> @gregw note also that you this PR is counting over the entries in the buckets multiple times.
> 
> Given that `QueuedPool` by default has `Integer.MAX_VALUE` capacity, a load spike may create a lot of entries and counting through them for filtering, for calculating sizes that are in a race with usages, etc. it is exactly the cost that now is causing CPU usage issues.
> 
> This PR still leaves buckets that are full of idle buffers in case of "adapts to changes in size frequency" possibly hitting exactly only on the largest bucket, it will consume a lot of CPU in worst cases due to running over possibly millions of entries (again, the worst case) to check whether they are idle, it is not able to respect maxMemory, all the checks for idleness and choice of buckets are racy, making it eventually impossible to understand bad behaviors that we may have not discovered yet.

I think you might be overstating the worst case. If it can't find idle entries then it ain't going to be full.

We will never have millions of entries to iterate over. Max size of concurrent pool is 256. If they are all in use then the queue is full of idle entries.

However, I'm definitely open to improving the algorithm, but it must be an adaptive one.

Reply:
I've added an optimization to reserve, so that if the size is at the maximum and the current bucket already has some idle entries, then the buffer is not added to the pool (just like in #11391).

I've also added optimized implementations for lots of getters on `CompoundPool` and `QueuedPool`. Specifically the `QueuedPool` only contains idle entries, so no need to stream to count them, we can just return the size.   This avoids a lot of iteration.

Also the the ConcurrentPool has some non-default implementations to avoid using the inefficient stream based implementation.